File: modules/slicot/src/fortran/MB03FD.f
      SUBROUTINE MB03FD( N, PREC, A, LDA, B, LDB, Q1, LDQ1, Q2, LDQ2,
     $                   DWORK, LDWORK, INFO )
C
C     PURPOSE
C
C     To compute orthogonal matrices Q1 and Q2 for a real 2-by-2 or
C     4-by-4 regular pencil
C
C                   ( A11  0  )     (  0  B12 )
C       aA - bB = a (         ) - b (         ),                     (1)
C                   (  0  A22 )     ( B21  0  )
C
C     such that Q2' A Q1 is upper triangular, Q2' B Q1 is upper quasi-
C     triangular, and the eigenvalues with negative real parts (if there
C     are any) are allocated on the top. The notation M' denotes the
C     transpose of the matrix M. The submatrices A11, A22, and B12 are
C     upper triangular. If B21 is 2-by-2, then all the other blocks are
C                                    -1        -1
C     nonsingular and the product A11   B12 A22   B21 has a pair of
C     complex conjugate eigenvalues.
C
C     ARGUMENTS
C
C     Input/Output Parameters
C
C     N       (input) INTEGER
C             The order of the input pencil, N = 2 or N = 4.
C
C     PREC    (input) DOUBLE PRECISION
C             The machine precision, (relative machine precision)*base.
C             See the LAPACK Library routine DLAMCH.
C
C     A       (input/output) DOUBLE PRECISION array, dimension (LDA, N)
C             On entry, the leading N-by-N part of this array must
C             contain the matrix A of the pencil aA - bB.
C             If N = 2, the diagonal elements only are referenced.
C             On exit, if N = 4, the leading N-by-N part of this array
C             contains the transformed upper triangular matrix of the
C             generalized real Schur form of the pencil aA - bB.
C             If N = 2, this array is unchanged on exit.
C
C     LDA     INTEGER
C             The leading dimension of the array A.  LDA >= N.
C
C     B       (input/output) DOUBLE PRECISION array, dimension (LDB, N)
C             On entry, the leading N-by-N part of this array must
C             contain the matrix B of the pencil aA - bB.
C             If N = 2, the anti-diagonal elements only are referenced.
C             On exit, if N = 4, the leading N-by-N part of this array
C             contains the transformed real Schur matrix of the
C             generalized real Schur form of the pencil aA - bB.
C             If N = 2, this array is unchanged on exit.
C
C     LDB     INTEGER
C             The leading dimension of the array B.  LDB >= N.
C
C     Q1      (output) DOUBLE PRECISION array, dimension (LDQ1, N)
C             The leading N-by-N part of this array contains the first
C             orthogonal transformation matrix.
C
C     LDQ1    INTEGER
C             The leading dimension of the array Q1.  LDQ1 >= N.
C
C     Q2      (output) DOUBLE PRECISION array, dimension (LDQ2, N)
C             The leading N-by-N part of this array contains the second
C             orthogonal transformation matrix.
C
C     LDQ2    INTEGER
C             The leading dimension of the array Q2.  LDQ2 >= N.
C
C     Workspace
C
C     DWORK   DOUBLE PRECISION array, dimension (LDWORK)
C             If N = 2, then DWORK is not referenced.
C
C     LDWORK  INTEGER
C             The dimension of the array DWORK.
C             If N = 4, then LDWORK >= 63. For good performance LDWORK
C             should be generally larger.
C             If N = 2, then LDWORK >= 0.
C
C     Error Indicator
C
C     INFO    INTEGER
C             = 0: succesful exit;
C             = 1: the QZ iteration failed in the LAPACK routine DGGES;
C             = 2: another error occured during execution of DGGES.
C
C     METHOD
C
C     The algorithm uses orthogonal transformations as described on page
C     29 in [2].
C
C     REFERENCES
C
C     [1] Benner, P., Byers, R., Mehrmann, V. and Xu, H.
C         Numerical computation of deflating subspaces of skew-
C         Hamiltonian/Hamiltonian pencils.
C         SIAM J. Matrix Anal. Appl., 24 (1), pp. 165-190, 2002.
C
C     [2] Benner, P., Byers, R., Losse, P., Mehrmann, V. and Xu, H.
C         Numerical Solution of Real Skew-Hamiltonian/Hamiltonian
C         Eigenproblems.
C         Tech. Rep., Technical University Chemnitz, Germany,
C         Nov. 2007.
C
C     NUMERICAL ASPECTS
C
C     The algorithm is numerically backward stable.
C
C     CONTRIBUTOR
C
C     Matthias Voigt, Fakultaet fuer Mathematik, Technische Universitaet
C     Chemnitz, October 16, 2008.
C
C     REVISIONS
C
C     V. Sima, Aug. 2009 (SLICOT version of the routine MB03FD).
C     V. Sima, Oct. 2009, Nov. 2009, Oct. 2010, Nov. 2010, Mar. 2016,
C     Mai 2016.
C     M. Voigt, Jan. 2012.
C
C     KEYWORDS
C
C     Eigenvalue exchange, matrix pencil, upper (quasi-)triangular
C     matrix.
C
C     ******************************************************************
C
C     .. Parameters ..
      DOUBLE PRECISION   ZERO, ONE
      PARAMETER          ( ZERO = 0.0D+0, ONE = 1.0D+0 )
C
C     .. Scalar Arguments ..
      INTEGER            INFO, LDA, LDB, LDQ1, LDQ2, LDWORK, N
      DOUBLE PRECISION   PREC
C
C     .. Array Arguments ..
      DOUBLE PRECISION   A( LDA, * ), B( LDB, * ), DWORK( * ),
     $                   Q1( LDQ1, * ), Q2( LDQ2, * )
C
C     .. Local Scalars ..
      INTEGER            IDUM, IERR, IHI, ILO
      DOUBLE PRECISION   A11, A22, B12, B21, CO, SAFMIN, SCALA, SCALB,
     $                   SI, TMP
C
C     .. Local Arrays ..
      LOGICAL            BWORK( 4 )
      DOUBLE PRECISION   AS( 4, 4 ), BS( 4, 4 ), C( 4 ), R( 4 )
C
C     .. External Functions ..
      LOGICAL            SB02OW
      DOUBLE PRECISION   DLAMCH
      EXTERNAL           DLAMCH, SB02OW
C
C     .. External Subroutines ..
      EXTERNAL           DGGBAK, DGGES, DLACPY, DLARTG, MB04DL
C
C     .. Intrinsic Functions ..
      INTRINSIC          ABS, MAX, SIGN, SQRT
C
C     .. Executable Statements ..
C
C     For efficiency, the input arguments are not tested.
C
C     Computations.
C
      IF( N.EQ.4 ) THEN
C
C        Save A and B, since DGGES might not converge.
C
         CALL DLACPY( 'Full', N, N, A, LDA, AS, 4 )
         CALL DLACPY( 'Full', N, N, B, LDB, BS, 4 )
         CALL DGGES(  'Vector Computation', 'Vector Computation',
     $                'Sorted', SB02OW, N, B, LDB, A, LDA, IDUM, DWORK,
     $                DWORK( N+1 ), DWORK( 2*N+1 ), Q2, LDQ2, Q1, LDQ1,
     $                DWORK( 3*N+1 ), LDWORK-3*N, BWORK, INFO )
         IF( INFO.NE.0 ) THEN
C
C           Retry after balancing.
C
            CALL DLACPY( 'Full', N, N, AS, 4, A, LDA )
            CALL DLACPY( 'Full', N, N, BS, 4, B, LDB )
            CALL MB04DL( 'Both', N, ZERO, B, LDB, A, LDA, ILO, IHI, C,
     $                   R, DWORK, IDUM, IERR )
            CALL DGGES(  'Vector Computation', 'Vector Computation',
     $                   'Sorted', SB02OW, N, B, LDB, A, LDA, IDUM,
     $                   DWORK, DWORK( N+1 ), DWORK( 2*N+1 ), Q2, LDQ2,
     $                   Q1, LDQ1, DWORK( 3*N+1 ), LDWORK-3*N, BWORK,
     $                   IERR )
C
C           If DGGES fails again, error return based on previous call.
C
            IF( IERR.NE.0 ) THEN
               IF( INFO.GE.1 .AND. INFO.LE.4 ) THEN
                  INFO = 1
               ELSE
                  INFO = 2
               END IF
               RETURN
            END IF
            CALL DGGBAK( 'Both', 'Right', N, ILO, IHI, C, R, N, Q1,
     $                   LDQ1, INFO )
            CALL DGGBAK( 'Both', 'Left',  N, ILO, IHI, C, R, N, Q2,
     $                   LDQ2, INFO )
         END IF
      ELSE
         INFO = 0
C
C        Set Q1, and Q2 to I_2, or permuted I_2, if there are 0, purely
C        imaginary, or infinite eigenvalues.
C
         A11 = ABS( A( 1, 1 ) )
         A22 = ABS( A( 2, 2 ) )
         B21 = ABS( B( 2, 1 ) )
         B12 = ABS( B( 1, 2 ) )
C
         SAFMIN = DLAMCH( 'Safe minimum' )
         SCALA  = ONE / MAX( A11, A22, SAFMIN )
         SCALB  = ONE / MAX( B12, B21, SAFMIN )
C
         A11 = SCALA*A11
         A22 = SCALA*A22
         B21 = SCALB*B21
         B12 = SCALB*B12
         IF( A11.LE.PREC ) THEN
            Q1( 1, 1 ) = ONE
            Q1( 2, 1 ) = ZERO
            Q1( 1, 2 ) = ZERO
            Q1( 2, 2 ) = ONE
            Q2( 1, 1 ) = ZERO
            Q2( 2, 1 ) = ONE
            Q2( 1, 2 ) = ONE
            Q2( 2, 2 ) = ZERO
         ELSE IF( A22.LE.PREC ) THEN
            Q1( 1, 1 ) = ZERO
            Q1( 2, 1 ) = ONE
            Q1( 1, 2 ) = ONE
            Q1( 2, 2 ) = ZERO
            Q2( 1, 1 ) = ONE
            Q2( 2, 1 ) = ZERO
            Q2( 1, 2 ) = ZERO
            Q2( 2, 2 ) = ONE
         ELSE IF( B21.LE.PREC ) THEN
            Q1( 1, 1 ) = ONE
            Q1( 2, 1 ) = ZERO
            Q1( 1, 2 ) = ZERO
            Q1( 2, 2 ) = ONE
            Q2( 1, 1 ) = ONE
            Q2( 2, 1 ) = ZERO
            Q2( 1, 2 ) = ZERO
            Q2( 2, 2 ) = ONE
         ELSE IF( B12.LE.PREC ) THEN
            Q1( 1, 1 ) = ZERO
            Q1( 2, 1 ) = ONE
            Q1( 1, 2 ) = ONE
            Q1( 2, 2 ) = ZERO
            Q2( 1, 1 ) = ZERO
            Q2( 2, 1 ) = ONE
            Q2( 1, 2 ) = ONE
            Q2( 2, 2 ) = ZERO
         ELSE
            IF( SIGN( ONE, A( 1, 1 ) )*SIGN( ONE, A( 2, 2 ) )*
     $          SIGN( ONE, B( 2, 1 ) )*SIGN( ONE, B( 1, 2 ) ).GT.ZERO
     $        ) THEN
C
C              The pencil has two real eigenvalues.
C
               CALL DLARTG( SIGN( ONE, A( 1, 1 )*A( 2, 2 ) )*
     $                      SQRT( A22*B12 ), SQRT( A11*B21 ), CO, SI,
     $                      TMP )
               Q1( 1, 1 ) =  CO
               Q1( 2, 1 ) = -SI
               Q1( 1, 2 ) =  SI
               Q1( 2, 2 ) =  CO
               CALL DLARTG( SQRT( A11*B12 ), SQRT( A22*B21 ), CO, SI,
     $                      TMP )
               Q2( 1, 1 ) =  CO
               Q2( 2, 1 ) = -SI
               Q2( 1, 2 ) =  SI
               Q2( 2, 2 ) =  CO
            ELSE
               Q1( 1, 1 ) = ONE
               Q1( 2, 1 ) = ZERO
               Q1( 1, 2 ) = ZERO
               Q1( 2, 2 ) = ONE
               Q2( 1, 1 ) = ONE
               Q2( 2, 1 ) = ZERO
               Q2( 1, 2 ) = ZERO
               Q2( 2, 2 ) = ONE
            END IF
         END IF
      END IF
C
      RETURN
C *** Last line of MB03FD ***
      END

File: modules/slicot/src/c/MB03FD.c
/* Translated by Nelson f2c (version 20170901).
   You must link the resulting object file with the libraries:
    -lnlsf2c -lm   (in that order)
*/

#include "nelson_f2c.h"

/* Table of constant values */

static doublereal c_b5 = 1.;

EXPORTSYMBOL /* Subroutine */ int mb03fd_(
    n, prec, a, lda, b, ldb, q1, ldq1, q2, ldq2, dwork, ldwork, info) integer* n;
doublereal *prec, *a;
integer* lda;
doublereal* b;
integer* ldb;
doublereal* q1;
integer* ldq1;
doublereal* q2;
integer* ldq2;
doublereal* dwork;
integer *ldwork, *info;
{
    /* System generated locals */
    integer a_dim1, a_offset, b_dim1, b_offset, q1_dim1, q1_offset, q2_dim1, q2_offset, i__1;
    doublereal d__1, d__2, d__3;
    /* Builtin functions */
    double d_sign(), sqrt();
    /* Local variables */
    static integer idum;
    extern /* Subroutine */ int dgges_();
    static logical compg;
    extern logical sb02ow_();
    static logical bwork[4];
    static doublereal a11, a22, b12, b21, co, si;
    extern /* Subroutine */ int dlartg_();
    static doublereal tmp;
    /*     SLICOT RELEASE 5.0. */
    /*     Copyright (c) 2002-2010 NICONET e.V. */
    /*     This program is free software: you can redistribute it and/or */
    /*     modify it under the terms of the GNU General Public License as */
    /*     published by the Free Software Foundation, either version 2 of */
    /*     the License, or (at your option) any later version. */
    /*     This program is distributed in the hope that it will be useful, */
    /*     but WITHOUT ANY WARRANTY; without even the implied warranty of */
    /*     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the */
    /*     GNU General Public License for more details. */
    /*     You should have received a copy of the GNU General Public License */
    /*     along with this program.  If not, see */
    /*     <http://www.gnu.org/licenses/>. */
    /*     PURPOSE */
    /*     To compute orthogonal matrices Q1 and Q2 for a real 2-by-2 or */
    /*     4-by-4 regular pencil */
    /*                   ( A11  0  )     (  0  B12 ) */
    /*       aA - bB = a (         ) - b (         ),                     (1) */
    /*                   (  0  A22 )     ( B21  0  ) */
    /*     such that Q2' A Q1 is upper triangular, Q2' B Q1 is upper quasi- */
    /*     triangular, and the eigenvalues with negative real parts (if there */
    /*     are any) are allocated on the top. The submatrices A11, A22, and */
    /*     B12 are upper triangular. If B21 is 2-by-2, then all the other */
    /*     blocks are nonsingular and the product */
    /*        -1        -1 */
    /*     A11   B12 A22   B21 has a pair of complex conjugate eigenvalues. */
    /*     ARGUMENTS */
    /*     Input/Output Parameters */
    /*     N       (input) INTEGER */
    /*             The order of the input pencil, N = 2 or N = 4. */
    /*     PREC    (input) DOUBLE PRECISION */
    /*             The machine precision, (relative machine precision)*base. */
    /*             See the LAPACK Library routine DLAMCH. */
    /*     A       (input/output) DOUBLE PRECISION array, dimension (LDA, N) */
    /*             On entry, the leading N-by-N part of this array must */
    /*             contain the matrix A of the pencil aA - bB. */
    /*             If N = 2, the diagonal elements only are referenced. */
    /*             On exit, if N = 4, the leading N-by-N part of this array */
    /*             contains the transformed upper triangular matrix of the */
    /*             generalized real Schur form of the pencil aA - bB. */
    /*             If N = 2, this array is unchanged on exit. */
    /*     LDA     INTEGER */
    /*             The leading dimension of the array A.  LDA >= N. */
    /*     B       (input/output) DOUBLE PRECISION array, dimension (LDB, N) */
    /*             On entry, the leading N-by-N part of this array must */
    /*             contain the matrix B of the pencil aA - bB. */
    /*             If N = 2, the anti-diagonal elements only are referenced. */
    /*             On exit, if N = 4, the leading N-by-N part of this array */
    /*             contains the transformed real Schur matrix of the */
    /*             generalized real Schur form of the pencil aA - bB. */
    /*             If N = 2, this array is unchanged on exit. */
    /*     LDB     INTEGER */
    /*             The leading dimension of the array B.  LDB >= N. */
    /*     Q1      (output) DOUBLE PRECISION array, dimension (LDQ1, N) */
    /*             The leading N-by-N part of this array contains the first */
    /*             orthogonal transformation matrix. */
    /*     LDQ1    INTEGER */
    /*             The leading dimension of the array Q1.  LDQ1 >= N. */
    /*     Q2      (output) DOUBLE PRECISION array, dimension (LDQ2, N) */
    /*             The leading N-by-N part of this array contains the second */
    /*             orthogonal transformation matrix. */
    /*     LDQ2    INTEGER */
    /*             The leading dimension of the array Q2.  LDQ2 >= N. */
    /*     Workspace */
    /*     DWORK   DOUBLE PRECISION array, dimension (LDWORK) */
    /*             If N = 2, then DWORK is not referenced. */
    /*     LDWORK  INTEGER */
    /*             The dimension of the array DWORK. */
    /*             If N = 4, then LDWORK >= 63. For good performance LDWORK */
    /*             should be generally larger. */
    /*             If N = 2, then LDWORK >= 0. */
    /*     Error Indicator */
    /*     INFO    INTEGER */
    /*             = 0: succesful exit; */
    /*             = 1: the QZ iteration failed in the LAPACK routine DGGES; */
    /*             = 2: another error occured during execution of DGGES. */
    /*     METHOD */
    /*     The algorithm uses orthogonal transformations as described on page */
    /*     29 in [2]. */
    /*     REFERENCES */
    /*     [1] Benner, P., Byers, R., Mehrmann, V. and Xu, H. */
    /*         Numerical computation of deflating subspaces of skew- */
    /*         Hamiltonian/Hamiltonian pencils. */
    /*         SIAM J. Matrix Anal. Appl., 24 (1), pp. 165-190, 2002. */
    /*     [2] Benner, P., Byers, R., Losse, P., Mehrmann, V. and Xu, H. */
    /*         Numerical Solution of Real Skew-Hamiltonian/Hamiltonian */
    /*         Eigenproblems. */
    /*         Tech. Rep., Technical University Chemnitz, Germany, */
    /*         Nov. 2007. */
    /*     NUMERICAL ASPECTS */
    /*     The algorithm is numerically backward stable. */
    /*     CONTRIBUTOR */
    /*     Matthias Voigt, Fakultaet fuer Mathematik, Technische Universitaet */
    /*     Chemnitz, October 16, 2008. */
    /*     REVISIONS */
    /*     V. Sima, Aug. 2009 (SLICOT version of the routine MB03FD). */
    /*     V. Sima, Oct. 2009, Nov. 2009, Oct. 2010, Nov. 2010. */
    /*     KEYWORDS */
    /*     Eigenvalue exchange, matrix pencil, upper (quasi-)triangular */
    /*     matrix. */
    /*     ****************************************************************** */
    /*     .. Parameters .. */
    /*     .. Scalar Arguments .. */
    /*     .. Array Arguments .. */
    /*     .. Local Scalars .. */
    /*     .. Local Arrays .. */
    /*     .. External Functions .. */
    /*     .. External Subroutines .. */
    /*     .. Intrinsic Functions .. */
    /*     .. Executable Statements .. */
    /*     For efficiency, the input arguments are not tested. */
    /* Parameter adjustments */
    a_dim1 = *lda;
    a_offset = a_dim1 + 1;
    a -= a_offset;
    b_dim1 = *ldb;
    b_offset = b_dim1 + 1;
    b -= b_offset;
    q1_dim1 = *ldq1;
    q1_offset = q1_dim1 + 1;
    q1 -= q1_offset;
    q2_dim1 = *ldq2;
    q2_offset = q2_dim1 + 1;
    q2 -= q2_offset;
    --dwork;
    /* Function Body */
    *info = 0;
    /*     Computations. */
    if (*n == 4) {
        i__1 = *ldwork - *n * 3;
        dgges_("Vector Computation", "Vector Computation", "Sorted", sb02ow_, n, &b[b_offset], ldb,
            &a[a_offset], lda, &idum, &dwork[1], &dwork[*n + 1], &dwork[(*n << 1) + 1],
            &q2[q2_offset], ldq2, &q1[q1_offset], ldq1, &dwork[*n * 3 + 1], &i__1, bwork, info, 18L,
            18L, 6L);
        if (*info != 0) {
            if (*info >= 1 && *info <= 4) {
                *info = 1;
            } else if (*info != 6) {
                *info = 2;
            } else {
                *info = 0;
            }
        }
        return 0;
    } else {
        /*        The pencil has infinite eigenvalues. The code decides this when */
        /*        A is (numerically) singular. */
        a11 = (d__1 = a[a_dim1 + 1], abs(d__1));
        a22 = (d__1 = a[(a_dim1 << 1) + 2], abs(d__1));
        b21 = (d__1 = b[b_dim1 + 2], abs(d__1));
        b12 = (d__1 = b[(b_dim1 << 1) + 1], abs(d__1));
        compg = FALSE_;
        if (a11 <= *prec * a22) {
            q1[q1_dim1 + 1] = 1.;
            q1[q1_dim1 + 2] = 0.;
            q1[(q1_dim1 << 1) + 1] = 0.;
            q1[(q1_dim1 << 1) + 2] = 1.;
            q2[q2_dim1 + 1] = 0.;
            q2[q2_dim1 + 2] = 1.;
            q2[(q2_dim1 << 1) + 1] = 1.;
            q2[(q2_dim1 << 1) + 2] = 0.;
        } else if (a22 <= *prec * a11) {
            q1[q1_dim1 + 1] = 0.;
            q1[q1_dim1 + 2] = 1.;
            q1[(q1_dim1 << 1) + 1] = 1.;
            q1[(q1_dim1 << 1) + 2] = 0.;
            q2[q2_dim1 + 1] = 1.;
            q2[q2_dim1 + 2] = 0.;
            q2[(q2_dim1 << 1) + 1] = 0.;
            q2[(q2_dim1 << 1) + 2] = 1.;
        } else {
            compg = TRUE_;
        }
        if (compg) {
            /*           The pencil has a double zero eigenvalue. */
            if (b21 <= *prec * b12) {
                q1[q1_dim1 + 1] = 1.;
                q1[q1_dim1 + 2] = 0.;
                q1[(q1_dim1 << 1) + 1] = 0.;
                q1[(q1_dim1 << 1) + 2] = 1.;
                q2[q2_dim1 + 1] = 1.;
                q2[q2_dim1 + 2] = 0.;
                q2[(q2_dim1 << 1) + 1] = 0.;
                q2[(q2_dim1 << 1) + 2] = 1.;
            } else if (b12 <= *prec * b21) {
                q1[q1_dim1 + 1] = 0.;
                q1[q1_dim1 + 2] = 1.;
                q1[(q1_dim1 << 1) + 1] = 1.;
                q1[(q1_dim1 << 1) + 2] = 0.;
                q2[q2_dim1 + 1] = 0.;
                q2[q2_dim1 + 2] = 1.;
                q2[(q2_dim1 << 1) + 1] = 1.;
                q2[(q2_dim1 << 1) + 2] = 0.;
            } else {
                compg = TRUE_;
            }
        }
        if (compg) {
            /*           The pencil has real eigenvalues. */
            d__2 = a[a_dim1 + 1] * a[(a_dim1 << 1) + 2];
            d__1 = d_sign(&c_b5, &d__2) * sqrt(a22 * b12);
            d__3 = sqrt(a11 * b21);
            dlartg_(&d__1, &d__3, &co, &si, &tmp);
            q1[q1_dim1 + 1] = co;
            q1[q1_dim1 + 2] = -si;
            q1[(q1_dim1 << 1) + 1] = si;
            q1[(q1_dim1 << 1) + 2] = co;
            d__1 = sqrt(a11 * b12);
            d__2 = sqrt(a22 * b21);
            dlartg_(&d__1, &d__2, &co, &si, &tmp);
            q2[q2_dim1 + 1] = co;
            q2[q2_dim1 + 2] = -si;
            q2[(q2_dim1 << 1) + 1] = si;
            q2[(q2_dim1 << 1) + 2] = co;
        }
    }
    return 0;
    /* *** Last line of MB03FD *** */
} /* mb03fd_ */
